xarray/core/dataset.py

```diff
@@ -2322,6 +2323,22 @@ def interp_like(self, other, method='linear', assume_sorted=False,
             ds = self.reindex(object_coords)
         return ds.interp(numeric_coords, method, assume_sorted, kwargs)
 
+    # Helper method for sel()
+    def _maybe_cast_floats(self, indexers):
```

Comment:
Does anyone have a prior on the likely performance impact of this? We're creating a copy of the indexers on each `.sel` call. 

If a) the requirement to coerce is rare enough, and b) it's expensive to coerce, we could loop through and check whether there are any incongruencies, and only then coerce.

Is there a viable alternative in having the indexer do these casts?

Reply:
@max-sixty Thank you for your time and note. I modified it to avoid unnecessary looping or copying.

In my understanding, the root of the problem is that pandas index is float 64. Despite that the issue only mentions float32, I tried to generalize the solution, by casting to the coords type for any kind of float.

Reply:
Thanks @HasanAhmadQ7 !

I'll let someone who knows the indexing machinery review this next (@pydata/xarray), given how central this code path is and my relative knowledge of this. Thanks for your patience